spec/java/util/ArrayList.lsl

```diff
@@ -0,0 +1,1205 @@
+        //        aListSubtracting: array<any>,
+        //        aListDeductible: array<any>
+        //    ): list<any>;
+
+        //But Collection isn't List.
+        //But it has method: toArray();
+        //Maybe we must have two arrays as params ?
+
+        //storage = action SUBTRACTING_SETS(storage, c);
+
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun retainAll (c: Collection): boolean
+    {
+        // #problem.
+        //I want to create such action:
+
+        //define action INTERSECTION(
+        //        aListSubtracting: array<any>,
+        //        aListDeductible: array<any>
+        //    ): list<any>;
+
+        //But Collection isn't List.
+        //But it has method: toArray();
+        //Maybe we must have two arrays as params ?
+
+        //storage = action INTERSECTION(storage, c);
+
+        action NOT_IMPLEMENTED();
+    }
+
+
+    @private
+    @throws(["java.io.IOException"])
+    fun writeObject (s: ObjectOutputStream): void
+    {
+        var expectedModCount = modCount;
+
+        // #problem
+        //Here is cycle and some operations with ObjectOutputStream.
+        action NOT_IMPLEMENTED();
+
+        if (modCount != expectedModCount)
+        {
+            action THROW_NEW("java.util.ConcurrentModificationException", []);
+        }
+    }
+
+
+    @private
+    @throws(["java.io.IOException", "java.lang.ClassNotFoundException"])
+    fun readObject (s: ObjectInputStream): void
+    {
+        // #problem
+        //Here is cycle and some operations with ObjectInputStream.
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun listIterator (index: int): ListIterator
+    {
+        _rangeCheckForAdd(index, length);
+
+        result = new ListItr(state=Created,
+            cursor=index,
+            expectedModCount=modCount);
+    }
+
+
+    fun listIterator (): ListIterator
+    {
+        result = new ListItr(state=Created,
+            cursor=0,
+            expectedModCount=modCount);
+    }
+
+
+    fun iterator (): Iterator
+    {
+        result = new Itr(state=Created,
+            expectedModCount=modCount);
+    }
+
+
+    fun subList (fromIndex: int, toIndex: int): List
+    {
+        _subListRangeCheck(fromIndex, toIndex, length);
+
+        // #problem
+        //We don't have decision about sublists.
+        //TODO
+        result = new SubList(state=Created,
+            startIndex=fromIndex,
+            endIndex=toIndex);
+    }
+
+
+    fun forEach (action: Consumer): void
+    {
+        // #problem
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun spliterator (): Spliterator
+    {
+        result = new ArrayListSpliterator(state=Initialized,
+            //This is right ? "parent=self"
+            parent=self,
+            origin = 0,
+            est=-1,
+            expectedModCount=0);
+    }
+
+
+    fun removeIf (filter: Predicate): boolean
+    {
+        // #problem
+        //We don't know how to work with lambda.
```

Comment:
Incorrect. We do have an action `CALL` that is designed to make calls to functional Java interfaces (i.e. where only a single abstract method resides).

Reply:
Fixed

Reply:
My bad. If we are removing from the *whole list* of stored items, then yes, the loop is unavoidable until we would have LibSL-level lambdas and loop-like action.